### ISSUE TITLE
opening file mentioned in feed doctype

### ISSUE THREAD

**Issue by johniez** [kurtmckee/feedparser#107]
I am currently getting an "Unknown IO error" printed to stderr while using `feedparser.parse('http://feeds.feedburner.com/news_trailbusterscom?format=xml')`
It defines a header:
```
<?xml version="1.0" encoding="UTF-8"?>
<?xml-stylesheet type="text/xsl" media="screen" href="/~d/styles/rss2full.xsl"?><?xml-stylesheet type="text/css" media="screen" href="http://feeds.feedburner.com/~d/styles/itemcontent.css"?><!DOCTYPE rss PUBLIC "-//Netscape Communications//DTD RSS 0.91//EN" "http://my.netscape.com/publish/formats/rss-0.91.dtd">
```

I have been using strace to see where it is happening and I saw a `stat('http://my.netscape.com/publish/formats/rss-0.91.dtd')` call for the doctype from the xml header. I tried to parse a feed with doctype changed to `file://etc/hosts` and strace have disclosed a successful `stat()` and `open()` for the file a filled in the doctype url.

This behaviour seems a little bit suspicious to me. Allowing user input to open a file in the system is not much pretty.
Is this OK?

**Comment by twm** [comment]
No, this behavior is not okay and actually it is pretty serious. Perhaps feedparser should use defusedxml, which wraps a number of Python XML libraries to prevent this stuff, and has nice explanations of these vulnerabilities:

* [external entity expansion (remote)](https://pypi.python.org/pypi/defusedxml#external-entity-expansion-remote)
* [external entity expansion (local file)](https://pypi.python.org/pypi/defusedxml#external-entity-expansion-local-file)

**Comment by kurtmckee** [comment]
@johniez thanks for reporting this!

@twm, great suggestion! I'd like feedparser to be far more stable and secure than it is, so this may be a necessary change to protect users! I'll look into it as soon as I can!

**Comment by dmoklaf** [comment]
Any update? This looks like a serious vulnerability

**Comment by rgeronimi** [comment]
Any opinion as to whether this vulnerability shall be resolved or not?

**Comment by genevieve-me** [comment]
I believe this issue may actually be closed. It was originally reported in 2017, but starting in Python 3.7.1, external entity expansion was forbidden in `xml.sax`. The vulnerabilities page in the stdlib docs says that it is vulnerable to DoS attacks, specifically billion laughs, quadratic blowup, and large tokens if libexpat <2.6 (but I believe all major linux distros have either updated to 2.6 or backported the security fix).

It could be worth keeping this issue open to track those two DoS attacks, but I don't believe that there's potential to make feedparser access arbitrary files or execute code on any supported Python version.

**Comment by rgeronimi** [comment]
Indeed the documentation you have linked to confirms this. Thanks!
